### ISSUE TITLE
[1.0] Moving away from JavaScript implementation

### ISSUE THREAD

**Issue by thesayyn** [thesayyn/protoc-gen-ts#255]
## What is happening with Rust implementation 

In summary, protoc-gen-ts continues to generate TypeScript code from proto files. However, there has been a shift from TypeScript to Rust in the project. This change is explained in more detail in the accompanying issue for the switch, citing reasons for the transition.

### Implementation of protoc-gen-ts in JavaScript

- javascript was hard to maintain and test due to slowness of tsc
- the `typescript` package a  peer dependency of protoc-gen-ts which broke the plugin if a wrong version was used. See #225  #86
- generation speed of protoc-gen-ts was affected by NodeJS runtime performance. 
- depending on the typescript ast limited us is number of ways
- it was impossible to get concurrency into the plugin. 
- javascript implementation did not pass the protobuf conformance tests.
- typescript wasn't truly type safe

### Implementation of protoc-gen-ts in Rust

- concurrency built into the plugin
- utilizing swc/ast APIs instead of tsc/ast APIs.
- blazingly fast and easy to test
- successfully passes conformance tests
- truly type safe
- capable of emitting both JS and TS thanks to swc.

 > [!NOTE] 
 > the [plugin written is javascript](https://www.npmjs.com/package/protoc-gen-ts) is still the latest version published in npm, and it will stay that way for a while to avoid breaking people's code. 
 > you can try out the[ protoc-gen-ts written in rust](https://github.com/thesayyn/protoc-gen-ts/pkgs/npm/protoc-gen-ts) by running `npm install @thesayyn/protoc-gen-ts@0.0.0-7a6a547`

> [!CAUTION] 
> However, the protoc-gen-ts using Rust as its programming language should not affect the users, they will continue installing the plugin `protoc-gen-ts`.
> But beware that protoc-gen-ts switching to Rust changed the emitted code in number of ways, please try the new plugin and report the issues back.

**Comment by thesayyn** [comment]
Rolling releases of protoc-gen-ts with Rust can be found here: https://github.com/thesayyn/protoc-gen-ts/pkgs/npm/protoc-gen-ts

**Comment by sanyatuning** [comment]
Hi! Can I use this without npm and node?

**Comment by thesayyn** [comment]
Yes, you should be able to, however, we don't have prebuilt binaries available anywhere, so you'd have clone and compile using cargo.

**Comment by loeffel-io** [comment]
hey @thesayyn,

is this a replacement for protoc-gen-js which is compatible with the protoc-gen-grpc-web plugin or a full replacement for protoc-gen-js and protoc-gen-grpc-web?

```bash
protoc order.proto \                                                                                                                                             
  --ts_out=import_style=...:. \ # <--
  --grpc-web_out=import_style=typescript,mode=grpcweb:.
```

**Comment by thesayyn** [comment]
Its a replacement for both protoc-gen-grpc-web and protoc-gen-js.

**Comment by thesayyn** [comment]
I haven't been able to pay too much attention to this repo last Q3, Q4 though.

**Comment by treebirg** [comment]
Hey @thesayyn, apologies if this is intentional, but the "rust version" package distribution does not seem to include the compiled wasm and js files created by wasm-pack. This causes the binary to fail when installed as instructed by this ticket's description:

> you can try out the[ protoc-gen-ts written in rust](https://github.com/thesayyn/protoc-gen-ts/pkgs/npm/protoc-gen-ts) by running npm install @thesayyn/protoc-gen-ts@0.0.0-7a6a547

I believe this is due to the .gitignore in the dist directory being picked up by npm during publishing of the package [1]. Explicitly including the dist dir via the `files` field in package.json should fix this. **Edit: this is incorrect. Nested .gitignore files take precedence over `files` in package.json.**

That being said, I think some other changes need to be made to the Github Actions workflow to make it build, as some crates are a bit old now. At least I haven't been able to install cargo-patch with the latest rust version (1.88 ATTOW). The following changes _should_ make it work:
- lock rust version to 1.79
- install cargo-patch with `--locked`

I can open a PR if easier (and if this all sounds sensible to you). 

[1] [see paragraph regarding .npmignore and .gitignore in npm docs](https://docs.npmjs.com/cli/v10/configuring-npm/package-json?v=true#files)

**Comment by thesayyn** [comment]
Yes, please do. i am not able to spend time on this repo anymore.
